File: src/chain_parallel.py
from dotenv import load_dotenv
load_dotenv()

from langchain_google_genai import ChatGoogleGenerativeAI
llm = ChatGoogleGenerativeAI(model="gemini-1.5-flash")

from langchain_core.output_parsers import StrOutputParser
from langchain.prompts import PromptTemplate
from langchain.callbacks.tracers import ConsoleCallbackHandler
from langchain_core.runnables import RunnableLambda, RunnableParallel


prompt1 = PromptTemplate.from_template("次の文章から趣味を抽出して1つの単語で回答してください。\n{user_input}")
chain1 = prompt1 | llm | StrOutputParser()
prompt2 = PromptTemplate.from_template("次の文章から職業を抽出して1つの単語で回答してください。\n{user_input}")
chain2 = prompt2 | llm | StrOutputParser()

prompt3 = PromptTemplate.from_template("次の記述を{language}に翻訳してください。\n{hobby}, {occupation}")
chain3 = prompt3 | llm | StrOutputParser()

chain = RunnableParallel(
    {"hobby": chain1, "occupation": chain2, "language": RunnableLambda(lambda x: "英語")}
) | chain3

print(chain.invoke(
    "普段はAIエンジニアをしていますが、休日には外に出て鳥を観察しに行き、撮影をします", 
    config={'callbacks': [ConsoleCallbackHandler()]},
))

File: js/chain_parallel.js
import { config } from "dotenv";
config();

import { ChatGoogleGenerativeAI } from "@langchain/google-genai";
const llm = new ChatGoogleGenerativeAI({ model: "gemini-1.5-flash" });

import { StringOutputParser } from "@langchain/core/output_parsers";
import { PromptTemplate } from "@langchain/core/prompts";
import { RunnableLambda, RunnableParallel } from "@langchain/core/runnables";


const prompt1 = PromptTemplate.fromTemplate(
    "次の文章から趣味を抽出して1つの単語で回答してください。\n{user_input}"
);
const chain1 = prompt1.pipe(llm).pipe(new StringOutputParser());

const prompt2 = PromptTemplate.fromTemplate(
    "次の文章から職業を抽出して1つの単語で回答してください。\n{user_input}"
);
const chain2 = prompt2.pipe(llm).pipe(new StringOutputParser());

const prompt3 = PromptTemplate.fromTemplate(
    "次の記述を{language}に翻訳してください。\n{hobby}, {occupation}"
);
const chain3 = prompt3.pipe(llm).pipe(new StringOutputParser());

const chain = RunnableParallel.from({
    hobby: chain1,
    occupation: chain2,
    language: RunnableLambda.from((x) => "英語"),
}).pipe(chain3);

const result = await chain.invoke(
    { user_input: "普段はAIエンジニアをしていますが、休日には外に出て鳥を観察しに行き、撮影をします" },
);

console.log(result);
